src/core/qgsgpkgflusher.cpp

```diff
@@ -0,0 +1,119 @@
+{
+    Q_OBJECT
+
+  public slots:
+    void scheduleFlush( const QString &filename );
+
+    void flush( const QString &filename );
+
+  private:
+    QMap<QString, QTimer *> mScheduledFlushes;
+};
+
+QgsGpkgFlusher::QgsGpkgFlusher( QgsProject *project )
+  : QObject()
+{
+  connect( project, &QgsProject::layersAdded, this, &QgsGpkgFlusher::onLayersAdded );
+  mFlusher = new Flusher();
+  mFlusher->moveToThread( &mFlusherThread );
+  connect( this, &QgsGpkgFlusher::requestFlush, mFlusher, &Flusher::scheduleFlush );
+  mFlusherThread.start();
+}
+
+QgsGpkgFlusher::~QgsGpkgFlusher()
+{
+  mFlusherThread.quit();
+  mFlusherThread.wait();
+}
+
+void QgsGpkgFlusher::onLayersAdded( const QList<QgsMapLayer *> layers )
+{
+  for ( QgsMapLayer *layer : layers )
+  {
+    QgsVectorLayer *vl = dynamic_cast<QgsVectorLayer *>( layer );
+    if ( vl && vl->dataProvider() )
+    {
+      QString dataSourceUri = vl->dataProvider()->dataSourceUri();
+      QString filePath = dataSourceUri.left( dataSourceUri.indexOf( '|' ) );
+      QFileInfo fi( filePath );
+      if ( fi.isFile() )
+      {
+        connect( vl, &QgsVectorLayer::committedFeaturesAdded, [this, filePath]() { emit requestFlush( filePath ); } );
+        connect( vl, &QgsVectorLayer::committedAttributesAdded, [this, filePath]() { emit requestFlush( filePath ); } );
+        connect( vl, &QgsVectorLayer::committedAttributesDeleted, [this, filePath]() { emit requestFlush( filePath ); } );
+        connect( vl, &QgsVectorLayer::committedGeometriesChanges, [this, filePath]() { emit requestFlush( filePath ); } );
+        connect( vl, &QgsVectorLayer::committedAttributeValuesChanges, [this, filePath]() { emit requestFlush( filePath ); } );
+        connect( vl, &QgsVectorLayer::committedFeaturesRemoved, [this, filePath]() { emit requestFlush( filePath ); } );
+      }
+    }
+  }
+}
+
+void Flusher::scheduleFlush( const QString &filename )
+{
+  if ( mScheduledFlushes.contains( filename ) )
+  {
+    mScheduledFlushes.value( filename )->start( 500 );
+  }
+  else
+  {
+    QTimer *timer = new QTimer();
+    connect( timer, &QTimer::timeout, this, [this, filename]() { flush( filename ); } );
+    timer->setSingleShot( true );
+    mScheduledFlushes.insert( filename, timer );
+    timer->start( 500 );
+  }
+}
+
+void Flusher::flush( const QString &filename )
+{
+  sqlite3_database_unique_ptr db;
+  int status = db.open_v2( filename, SQLITE_OPEN_READWRITE, nullptr );
+  if ( status != SQLITE_OK )
+  {
+    QgsMessageLog::logMessage( QObject::tr( "There was an error opening the database <b>%1</b>: %2" ).arg( filename, db.errorMessage() ) );
+  }
+
+  QString error;
+  db.exec( "PRAGMA wal_checkpoint;", error );
+
+  if ( error.isEmpty() )
+  {
+    delete mScheduledFlushes[filename];
+    mScheduledFlushes.remove( filename );
+  }
+  else
+  {
+    QgsMessageLog::logMessage( QObject::tr( "Could not flush database %1 (%3) " ).arg( filename, error ) );
+    mScheduledFlushes[filename]->start( 500 );
```

Comment:
Don't you think the log get's crazy when this happens? Or when does it happen? Only once - then this is fine- or may be the file blocked for a minute and then we receive 120 messages?

Reply:
Yes in theory you are right. However, I didn't manage to get this message even a single time so far.
And if this happens more than a single time we have a severe problem which I want to be reported, so I didn't spend time to fine tune this.